gadgets/src/smt/binary_smt.rs

```diff
@@ -0,0 +1,202 @@
+		let mut cur_idx = ScalarBits::from_scalar(&idx, TREE_DEPTH);
+		let mut cur_node = self.root.clone();
+
+		let need_proof = proof.is_some();
+		let mut proof_vec = Vec::<Scalar>::new();
+
+		for _i in 0..self.depth {
+			let k = cur_node.to_bytes();
+			let v = self.db.get(&k).unwrap();
+			if cur_idx.is_msb_set() {
+				// MSB is set, traverse to right subtree
+				cur_node = v.1;
+				if need_proof { proof_vec.push(v.0); }
+			} else {
+				// MSB is unset, traverse to left subtree
+				cur_node = v.0;
+				if need_proof { proof_vec.push(v.1); }
+			}
+			cur_idx.shl();
+		}
+
+		match proof {
+			Some(v) => {
+				v.extend_from_slice(&proof_vec);
+			}
+			None => ()
+		}
+
+		cur_node
+	}
+
+	/// Verify a merkle proof, if `root` is None, use the current root else use given root
+	pub fn verify_proof(&self, idx: Scalar, val: Scalar, proof: &[Scalar], root: Option<&Scalar>) -> bool {
+		let mut cur_idx = ScalarBits::from_scalar(&idx, TREE_DEPTH);
+		let mut cur_val = val.clone();
+
+		for i in 0..self.depth {
+			cur_val = {
+				if cur_idx.is_lsb_set() {
+					// mimc(&proof[self.depth-1-i], &cur_val, self.hash_constants)
+					Poseidon_hash_2(proof[self.depth-1-i].clone(), cur_val.clone(), self.hash_params)
+				} else {
+					// mimc(&cur_val, &proof[self.depth-1-i], self.hash_constants)
+					Poseidon_hash_2(cur_val.clone(), proof[self.depth-1-i].clone(), self.hash_params)
+				}
+			};
+
+			cur_idx.shr();
+		}
+
+		// Check if root is equal to cur_val
+		match root {
+			Some(r) => {
+				cur_val == *r
+			}
+			None => {
+				cur_val == self.root
+			}
+		}
+	}
+
+	fn update_db_with_key_val(&mut self, key: Scalar, val: DBVal) {
+		self.db.insert(key.to_bytes(), val);
+	}
+}
+
+
+/// left = (1-leaf_side) * leaf + (leaf_side * proof_node)
+/// right = leaf_side * leaf + ((1-leaf_side) * proof_node))
+pub fn vanilla_merkle_merkle_tree_verif_gadget<CS: ConstraintSystem>(
+	cs: &mut CS,
+	depth: usize,
+	root: &Scalar,
+	leaf_val: AllocatedScalar,
+	leaf_index_bits: Vec<AllocatedScalar>,
+	proof_nodes: Vec<AllocatedScalar>,
+	statics: Vec<AllocatedScalar>,
+	poseidon_params: &Poseidon
+) -> Result<(), R1CSError> {
+
+	let mut prev_hash = LinearCombination::default();
+
+	let statics: Vec<LinearCombination> = statics.iter().map(|s| s.variable.into()).collect();
+
+	for i in 0..depth {
+		let leaf_val_lc = if i == 0 {
+			LinearCombination::from(leaf_val.variable)
+		} else {
+			prev_hash.clone()
+		};
+		let one_minus_leaf_side: LinearCombination = Variable::One() - leaf_index_bits[i].variable;
+
+		let (_, _, left_1) = cs.multiply(one_minus_leaf_side.clone(), leaf_val_lc.clone());
+		let (_, _, left_2) = cs.multiply(leaf_index_bits[i].variable.into(), proof_nodes[i].variable.into());
+		let left = left_1 + left_2;
+
+		let (_, _, right_1) = cs.multiply(leaf_index_bits[i].variable.into(), leaf_val_lc);
+		let (_, _, right_2) = cs.multiply(one_minus_leaf_side, proof_nodes[i].variable.into());
+		let right = right_1 + right_2;
```

Comment:
We are using a different method for figuring out the side: https://github.com/edgeware-builders/anon/blob/master/pallets/merkle/src/lib.rs#L463-L468